src/serializer/functions.js

```diff
@@ -16,91 +16,127 @@ import invariant from "../invariant.js";
 import { type Effects, type PropertyBindings, Realm } from "../realm.js";
 import type { PropertyBinding } from "../types.js";
 import { ignoreErrorsIn } from "../utils/errors.js";
-import { AbstractObjectValue, FunctionValue, ObjectValue } from "../values/index.js";
+import {
+  AbstractObjectValue,
+  FunctionValue,
+  ObjectValue,
+  type ECMAScriptSourceFunctionValue,
+} from "../values/index.js";
 import { ModuleTracer } from "./modules.js";
 import buildTemplate from "babel-template";
 import * as t from "babel-types";
 
 export class Functions {
-  constructor(realm: Realm, functions: ?Array<string>, moduleTracer: ModuleTracer) {
+  constructor(
+    realm: Realm,
+    functions: ?Array<string>,
+    moduleTracer: ModuleTracer,
+    runtimeFunctions: Set<ECMAScriptSourceFunctionValue, string>
+  ) {
     this.realm = realm;
     this.functions = functions;
     this.moduleTracer = moduleTracer;
     this.writeEffects = new Map();
-    this.nameToFunctionValue = new Map();
+    this.runtimeFunctions = runtimeFunctions;
+    this.functionToString = new Map();
   }
 
   realm: Realm;
   functions: ?Array<string>;
+  // maps back from FunctionValue to the expression string
+  functionToString: Map<FunctionValue, string>;
   moduleTracer: ModuleTracer;
-  writeEffects: Map<string, Effects>;
-  nameToFunctionValue: Map<string, FunctionValue>;
+  writeEffects: Map<FunctionValue, Effects>;
+  // additional functions specified with __registerAdditionalFunction at runtime
+  runtimeFunctions: Map<ECMAScriptSourceFunctionValue, string>;
 
   checkThatFunctionsAreIndependent() {
     let functions = this.functions;
-    invariant(functions, "This method should only be called if initialized with defined functions");
+    invariant(
+      functions || this.runtimeFunctions.size > 0,
+      "This method should only be called if initialized with defined functions"
+    );
 
     // lookup functions
     let calls = [];
-    for (let fname of functions) {
-      let fun;
-      let fnameAst = buildTemplate(fname)({}).expression;
-      if (fnameAst) {
-        try {
-          let e = ignoreErrorsIn(this.realm, () => this.realm.evaluateNodeForEffectsInGlobalEnv(fnameAst));
-          fun = e ? e[0] : undefined;
-        } catch (ex) {
-          if (!(ex instanceof ThrowCompletion)) throw ex;
+    if (functions) {
+      for (let fname of functions) {
```

Comment:
Instead of the additional `if (functions) {` nesting, could you just say `for (let fname of (functions || []) {`? Smaller change, less control-flow to understand.